cmd/ftl/cmd_dev.go

```diff
@@ -2,296 +2,45 @@ package main
-			logger.Debugf("Detected %d failed modules, previously had %d", len(failedModules), previousFailures)
-			for module := range failedModules {
-				modules.ForceRebuild(module)
-			}
-			previousFailures = len(failedModules)
-		}
-
-		if allModulesDeployed && d.ExitAfterDeploy {
-			logger.Infof("All modules deployed successfully.")
-			cancel()
-			return wg.Wait()
-		}
-
-		select {
-		case module := <-schemaChanges:
-			modules.RebuildDependentModules(ctx, module)
-
-		drainLoop: // Drain all messages from the channel to avoid extra redeploys
-			for {
-				select {
-				case module := <-schemaChanges:
-					modules.RebuildDependentModules(ctx, module)
-				default:
-					break drainLoop
-				}
-			}
-		case <-time.After(delay):
-		case <-ctx.Done():
-			return wg.Wait()
-		}
-	}
-}
-
-func (d *devCmd) watchForSchemaChanges(ctx context.Context, client ftlv1connect.ControllerServiceClient, schemaChanges chan *schema.Module) error {
-	logger := log.FromContext(ctx)
-
-	for {
-		select {
-		case <-ctx.Done():
-			return nil
-		default:
-		}
-
-		stream, err := client.PullSchema(ctx, connect.NewRequest(&ftlv1.PullSchemaRequest{}))
-		if err != nil {
-			return err
-		}
-
-		for stream.Receive() {
-			select {
-			case <-ctx.Done():
-				logger.Warnf("Context canceled during schema change streaming, closing stream...")
-				stream.Close()
-				return ctx.Err()
-			default:
-			}
-
-			msg := stream.Msg()
-			if msg.ChangeType == ftlv1.DeploymentChangeType_DEPLOYMENT_ADDED || msg.ChangeType == ftlv1.DeploymentChangeType_DEPLOYMENT_CHANGED {
-				module, err := schema.ModuleFromProto(msg.Schema)
-				if err != nil {
-					return err
-				}
-				schemaChanges <- module
-			}
-		}
-
-		if errors.Is(ctx.Err(), context.Canceled) {
-			logger.Infof("Stream disconnected, attempting to reconnect...")
-		}
-
-		stream.Close()
-		time.Sleep(d.ReconnectDelay)
-	}
-}
-
-func (d *devCmd) getTomls() ([]string, error) {
-	baseDir := d.BaseDir
-	tomls := []string{}
-
-	err := buildengine.WalkDir(baseDir, func(srcPath string, d fs.DirEntry) error {
-		if filepath.Base(srcPath) == "ftl.toml" {
-			tomls = append(tomls, srcPath)
-			return buildengine.ErrSkip // Return errSkip to stop recursion in this branch
-		}
-		return nil
-	})
+	err = d.ServeCmd.pollControllerOnine(ctx, client)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return tomls, nil
-}
-
-func (d *devCmd) addOrRemoveModules(tomls []string, modules moduleMap) error {
-	for _, toml := range tomls {
-		dir := filepath.Dir(toml)
-		if _, ok := modules[dir]; !ok {
-			config, err := moduleconfig.LoadModuleConfig(dir)
-			if err != nil {
-				return err
-			}
-			modules.AddModule(dir, config.Module)
-		}
-	}
+	g.Go(func() error {
+		return engine.Dev(ctx, d.Watch)
+	})
 
-	for dir := range modules {
-		found := false
-		for _, toml := range tomls {
-			if filepath.Dir(toml) == dir {
-				found = true
-				break
-			}
-		}
-		if !found {
-			modules.RemoveModule(dir)
-		}
-	}
-	return nil
```

Comment:
So much deleted code!

Reply:
It's all that sweet sweet `buildengine` magic now :)